File: PureWebUI/PureWebUI/OptionsViewController.m
//  OptionsViewController.m
//
//  Created by Chris Burns on 2/10/14.
//  Copyright 2014 Calgary Scientific Inc. All rights reserved.

#import "OptionsViewController.h"

#import <PureWeb/PureWeb.h>

#import "DiagnosticViewDelegate.h"
@interface OptionsViewController ()

@property (strong, nonatomic) IBOutlet UISlider *interactiveQualitySlider;
@property (strong, nonatomic) IBOutlet UISlider *fullQualitySlider;


@end

@implementation OptionsViewController

- (PWMutableEncoderConfiguration *)encoderConfiguration
{
    return [DiagnosticViewDelegate sharedInstance].encoderConfiguration;
}


- (void)viewDidLoad
{
    [self setup];
    [super viewDidLoad];
}

- (void)setup {
    
    [self setupPickerView:self.interactiveMimeTypePickerView];
    [self setupPickerView:self.fullQualityMimeTypePickerView];
    
    //Setup the Default Options Using the Configuration
    self.interactiveQualitySlider.value = [self encoderConfiguration].interactiveQuality.quality;
    self.fullQualitySlider.value = [self encoderConfiguration].fullQuality.quality;
    
    NSInteger interactiveRow = [self indexOfMimeType:[self encoderConfiguration].interactiveQuality.mimeType];
    [self.interactiveMimeTypePickerView selectRow:interactiveRow inComponent:0 animated:NO];

    NSInteger fullQualityRow = [self indexOfMimeType:[self encoderConfiguration].fullQuality.mimeType];
    [self.fullQualityMimeTypePickerView selectRow:fullQualityRow inComponent:0 animated:NO];
    
    //Setup PureWeb Event Listeners to Listen for Service Side Updates
    [self.encoderConfiguration.interactiveQuality.changed addSubscriber:self action:@selector(interactiveQualityDidChange:)];
    [self.encoderConfiguration.fullQuality.changed addSubscriber:self action:@selector(fullQualityDidChange:)];
}

- (void)setupPickerView:(UIPickerView *) pickerView {
    
    pickerView.dataSource = self;
    pickerView.delegate = self;
}

- (IBAction)interactiveQualitySliderValueChanged:(UISlider *)sender {
    
    self.encoderConfiguration.interactiveQuality.quality = (NSInteger) sender.value;
    
}

- (IBAction)fullQualitySliderValueChanged:(UISlider *)sender {
    
    self.encoderConfiguration.fullQuality.quality = (NSInteger) sender.value;
}

- (void)interactiveMimeTypeValueChanged:(NSString *) updatedMimeType
{
    self.encoderConfiguration.interactiveQuality.mimeType = updatedMimeType;

}

- (void)fullQualityMimeTypeValueChanged:(NSString *) updatedMimeType
{
    self.encoderConfiguration.fullQuality.mimeType = updatedMimeType;
}

#pragma mark PW Delegates

- (void)interactiveQualityDidChange:(NSObject *)object
{
    NSInteger interactiveRow = [self indexOfMimeType:[self encoderConfiguration].interactiveQuality.mimeType];
    [self.interactiveMimeTypePickerView selectRow:interactiveRow inComponent:0 animated:NO];
    
    self.interactiveQualitySlider.value = self.encoderConfiguration.interactiveQuality.quality;
}
- (void)fullQualityDidChange:(NSObject *)object
{
    NSInteger fullQualityRow = [self indexOfMimeType:[self encoderConfiguration].fullQuality.mimeType];
    [self.fullQualityMimeTypePickerView selectRow:fullQualityRow inComponent:0 animated:NO];
    
    self.fullQualitySlider.value = self.encoderConfiguration.fullQuality.quality;
}


#pragma mark Picker View Data Source Delegate
// returns the number of 'columns' to display.
- (NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView {
    return 1;
}

// returns the # of rows in each component..
- (NSInteger)pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component {
    return 3;
}

#pragma mark Picker View Delegate
- (NSArray *)mimeTypes
{
    static NSArray *mimeTypes;
    
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        mimeTypes = @[kSupportedEncoderMimeTypeTiles, kSupportedEncoderMimeTypeJpeg, kSupportedEncoderMimeTypePng];
    });
    
    return mimeTypes;
}

- (NSString *)mimeTypeAtIndex:(NSInteger)index
{
    return [self.mimeTypes objectAtIndex:index];
}

- (NSInteger)indexOfMimeType:(NSString *)mimeType
{
    return [self.mimeTypes indexOfObject:mimeType];
}

- (NSString *)pickerView:(UIPickerView *)pickerView titleForRow:(NSInteger)row forComponent:(NSInteger)component
{
    return [self mimeTypeAtIndex:row];
}

- (void)pickerView:(UIPickerView *)pickerView didSelectRow:(NSInteger)row inComponent:(NSInteger)component
{
    if([pickerView isEqual:self.interactiveMimeTypePickerView]) {
        
        [self interactiveMimeTypeValueChanged:[self mimeTypeAtIndex:row]];
    }
    else {
        [self fullQualityMimeTypeValueChanged:[self mimeTypeAtIndex:row]];
    }
    
}


@end

File: Swift-Asteroids/Asteroids/Asteroids/OptionsViewController.swift
//
//  OptionsViewController.swift
//  Asteroids
//
//  Created by Jonathan Neitz on 2016-02-16.
//  Copyright © 2016 Calgary Scientific Inc. All rights reserved.
//

import UIKit


class OptionsViewController: UIViewController, UIPickerViewDelegate, UIPickerViewDataSource {

    @IBOutlet weak var interactiveMimePicker: UIPickerView!
    
    @IBOutlet weak var nonInteractiveMimePicker: UIPickerView!
    
    @IBOutlet weak var interactiveQualitySlider: UISlider!
    
    @IBOutlet weak var nonInteractiveQualitySlider: UISlider!
    
    let divisor : Float = 100.0 // Conversion for the UISlider 0.0 to 1.0 to quality range between 0 and 100
    
    let mimeTypes = ["image/x-tile","image/jpeg","image/png"];
    
    override func viewDidLoad() {
        super.viewDidLoad()
        setup()
    }

    func setup() {
        interactiveMimePicker.delegate = self
        interactiveMimePicker.dataSource = self
        nonInteractiveMimePicker.delegate = self
        nonInteractiveMimePicker.dataSource = self
        
        let encoderConfiguration = DiagnosticViewDelegate.encoderConfiguration
        
        interactiveQualitySlider.value = Float(encoderConfiguration.interactiveQuality.quality) / divisor
        nonInteractiveQualitySlider.value = Float(encoderConfiguration.fullQuality.quality) / divisor
        
        setMimePickerRow(encoderConfiguration.interactiveQuality.mimeType, picker: interactiveMimePicker)
        setMimePickerRow(encoderConfiguration.fullQuality.mimeType, picker: nonInteractiveMimePicker)
        
        encoderConfiguration.interactiveQuality.changed.addSubscriber(self, action: #selector(OptionsViewController.interactiveQualityChanged(_:)))
        encoderConfiguration.fullQuality.changed.addSubscriber(self, action: #selector(OptionsViewController.nonInteractiveQualityChanged(_:)))
    }
    
    func setMimePickerRow(mimeType : String, picker : UIPickerView ) {
        var index : Int = 0
        for mime in mimeTypes {
            if mimeType == mime {
                picker.selectRow(index, inComponent: 0, animated: false)
            }
            
            index += 1;
        }
    }
    
    // MARK: UISlider actions
    
    @IBAction func interactiveSliderChanged(sender: UISlider) {
        DiagnosticViewDelegate.encoderConfiguration.interactiveQuality.quality = Int(sender.value * divisor)
    }
    
    @IBAction func nonInteractiveSliderChanged(sender: UISlider) {
        DiagnosticViewDelegate.encoderConfiguration.fullQuality.quality = Int(sender.value * divisor)
    }
    
    // MARK: UIPickerView actions
    
    func interactiveQualityChanged(object : NSObject) {
        let encoderConfiguration = DiagnosticViewDelegate.encoderConfiguration
        
        interactiveQualitySlider.value = Float(encoderConfiguration.interactiveQuality.quality) / divisor
        setMimePickerRow(encoderConfiguration.interactiveQuality.mimeType, picker: interactiveMimePicker)
        
    }
    
    func nonInteractiveQualityChanged(object: NSObject) {
        let encoderConfiguration = DiagnosticViewDelegate.encoderConfiguration
        
        nonInteractiveQualitySlider.value = Float(encoderConfiguration.fullQuality.quality) / divisor
        setMimePickerRow(encoderConfiguration.fullQuality.mimeType, picker: nonInteractiveMimePicker)
    }
    
    // MARK: - UIPickerView Delegate and Data source methods
    
    func numberOfComponentsInPickerView(pickerView: UIPickerView) -> Int {
        return 1
    }
    
    func pickerView(pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
        return 3;
    }

    func pickerView(pickerView: UIPickerView, titleForRow row: Int, forComponent component: Int) -> String? {
        return mimeTypes[row];
    }
    
    func pickerView(pickerView: UIPickerView, didSelectRow row: Int, inComponent component: Int) {
       
        let encoderConfiguration = DiagnosticViewDelegate.encoderConfiguration
        
        if pickerView == self.interactiveMimePicker {
            encoderConfiguration.interactiveQuality.mimeType = mimeTypes[row]
        } else {
            encoderConfiguration.fullQuality.mimeType = mimeTypes[row]
        }
    }
    
}
